mypy/subtypes.py

```diff
@@ -303,6 +331,227 @@ def visit_type_type(self, left: TypeType) -> bool:
         return False
 
 
+@contextmanager
+def pop_on_exit(stack: List[Tuple[Instance, Instance]],
+                left: Instance, right: Instance) -> Iterator[None]:
+    stack.append((left, right))
+    yield
+    stack.pop()
+
+
+def is_protocol_implementation(left: Instance, right: Instance, allow_any: bool = True) -> bool:
+    """Check whether 'left' implements the protocol 'right'. If 'allow_any' is False, then
+    check for a proper subtype. Treat recursive protocols by using the 'assuming'
+    structural subtype matrix (in sparse representation), see comment in nodes.TypeInfo.
+    """
+    assert right.type.is_protocol
+    assuming = right.type.assuming if allow_any else right.type.assuming_proper
+    for (l, r) in reversed(assuming):
+        if sametypes.is_same_type(l, left) and sametypes.is_same_type(r, right):
+            return True
+    with pop_on_exit(assuming, left, right):
+        if right.type.protocol_members is None:
+            # This type has not been yet analyzed, probably a call from make_simplified_union.
```

Comment:
Verify that this is still needed (there was a recent fix by Jukka that postponed calling ``make_simplified_union``/``join_type_list`` to a later stage). Open an issue if necessary.